File: leetcode/python/36.Valid Sudoku.py
class Main:

  def isValidSudoku(self,board):
    helper = set()
    # helper = {'a'}

    for i in range(len(board)):
      helper.clear()
      for j in range(len(board)):
        print(board[i][j],end=" ")
        if not self.process(helper,board[i][j]):
          return False
      print("\n")
    
    for j in range(len(board)):
      helper.clear()
      for i in range(len(board)):
        if not self.process(helper,board[i][j]):
          return False
    
    for i in range(0,9,3):
      for j in range(0,9,3):
        helper.clear()
        for k in range(9):
          if not self.process(helper,board[int(i+k/3)][int(j+k%3)]):
            return False
    return True
  
  def process(self,helper,c):
    if c == '.':
      return True
    if c in helper:
      return False
    helper.add(c)
    return True

m = Main()
c  = [
  ['5','3','.','.','7','.','.','.','.'],
  ['6','.','.','1','9','5','.','.','.'],
  ['.','9','8','.','.','.','.','6','.'],
  ['8','.','.','.','6','.','.','.','3'],
  ['4','.','.','8','.','3','.','.','1'],
  ['7','.','.','.','2','.','.','.','6'],
  ['.','6','.','.','.','.','2','8','.'],
  ['.','.','.','4','1','9','.','.','5'],
  ['.','.','.','.','8','.','.','7','9']
]
print(m.isValidSudoku(c))

File: leetcode/kotlin/36.Valid Sudoku.kt
object Main {
    private fun isValidSudoku(board: Array<CharArray>): Boolean {
        val helper: MutableSet<Char> = HashSet()
        for (i in board.indices) {
            helper.clear()
            for (j in board.indices) {
                print(board[i][j].toString() + " ")
                if (!process(helper, board[i][j])) return false
            }
            println("\n")
        }
        for (j in board.indices) {
            helper.clear()
            for (i in board.indices) {
                if (!process(helper, board[i][j])) return false
            }
        }
        var i = 0
        while (i < 9) {
            var j = 0
            while (j < 9) {
                helper.clear()
                for (k in 0..8) {
                    if (!process(helper, board[i + k / 3][j + k % 3])) return false
                }
                j += 3
            }
            i += 3
        }
        return true
    }

    private fun process(helper: MutableSet<Char>, c: Char): Boolean {
        if (c == '.') {
            return true
        }
        if (helper.contains(c)) {
            return false
        }
        helper.add(c)
        return true
    }

    @JvmStatic
    fun main(s: Array<String>) {
        val c = arrayOf(
            charArrayOf('5', '3', '.', '.', '7', '.', '.', '.', '.'),
            charArrayOf('6', '.', '.', '1', '9', '5', '.', '.', '.'),
            charArrayOf('.', '9', '8', '.', '.', '.', '.', '6', '.'),
            charArrayOf('8', '.', '.', '.', '6', '.', '.', '.', '3'),
            charArrayOf('4', '.', '.', '8', '.', '3', '.', '.', '1'),
            charArrayOf('7', '.', '.', '.', '2', '.', '.', '.', '6'),
            charArrayOf('.', '6', '.', '.', '.', '.', '2', '8', '.'),
            charArrayOf('.', '.', '.', '4', '1', '9', '.', '.', '5'),
            charArrayOf('.', '.', '.', '.', '8', '.', '.', '7', '9')
        )
        println(isValidSudoku(c))
    }
}